pkg/storageops/gce/gce.go

```diff
@@ -0,0 +1,540 @@
+// are set as env variables
+func IsDevMode() bool {
+	var i = new(instance)
+	gceInfoFromEnv(i)
+	return len(i.Error) == 0
+}
+
+// NewClient creates a new GCE operations client
+func NewClient() (storageops.Ops, error) {
+	var i = new(instance)
+	if metadata.OnGCE() {
+		gceInfo(i)
+	} else if ok := IsDevMode(); ok {
+		gceInfoFromEnv(i)
+	} else {
+		return nil, fmt.Errorf("instance is not running on GCE")
+	}
+
+	if len(i.Error) != 0 {
+		return nil, fmt.Errorf("error while fetching instance information. Err: %v", i.Error)
+	}
+
+	c, err := google.DefaultClient(context.Background(), compute.ComputeScope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to authenticate with google api. Err: %v", err)
+	}
+
+	service, err := compute.New(c)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Compute service: %v", err)
+	}
+
+	return &gceOps{
+		inst:    i,
+		service: service,
+	}, nil
+}
+
+func (s *gceOps) waitStatus(id string, desired string) error {
+	actual := ""
+	for retries, maxRetries := 0, storageops.ProviderOpsMaxRetries; actual != desired && retries < maxRetries; retries++ {
+		d, err := s.service.Disks.Get(s.inst.Project, s.inst.Zone, id).Do()
+		if err != nil {
+			if ignore := notFoundRegex.MatchString(err.Error()); ignore {
+				time.Sleep(storageops.ProviderOpsRetryInterval)
+				continue
+			}
+
+			return err
+		}
+
+		if d == nil {
+			return fmt.Errorf("expected one disk %v got none", id)
+		}
+
+		actual = d.Status
+		if len(actual) == 0 {
+			return fmt.Errorf("nil volume state for %v", id)
+		}
+
+		if actual == desired {
+			break
+		}
+
+		time.Sleep(storageops.ProviderOpsRetryInterval)
+	}
+
+	if actual != desired {
+		return fmt.Errorf("disk %v did not transition to %v current state %v",
+			id, desired, actual)
+	}
+
+	return nil
+}
+
+func (s *gceOps) waitForDetach(
+	diskURL string,
+	timeout time.Duration,
+) (err error) {
+	var inst *compute.Instance
+
+	interval := 5 * time.Second
+	for elapsed := 0 * time.Second; elapsed < timeout; elapsed += interval {
+		inst, err = s.describeinstance()
+		if err != nil {
+			return
+		}
+
+		found := false
+		for _, d := range inst.Disks {
+			if d.Source == diskURL {
+				found = true
+				break
+			}
+		}
+
+		if found {
+			time.Sleep(interval)
+			continue
+		} else {
+			return
+		}
+	}
+
+	err = fmt.Errorf("disk: %s is still attached to instance: %s", diskURL, s.inst.Name)
+	return
+}
+
+// waitForAttach checks if given disk is attached to the local instance
+func (s *gceOps) waitForAttach(
+	disk *compute.Disk,
+	timeout time.Duration,
+) (devicePath string, err error) {
+	interval := 2 * time.Second
+	for elapsed := 0 * time.Second; elapsed < timeout; elapsed += interval {
+		devicePath, err = s.DevicePath(disk.Name)
+		if err == nil {
+			return
+		}
+
+		time.Sleep(interval)
+	}
+
+	err = fmt.Errorf("disk: %s is not attached to instance: %s", disk.Name, s.inst.Name)
+	return
+}
+
+func (s *gceOps) Name() string { return "gce" }
```

Comment:
Do you mind grouping the exposed functions?

Reply:
done